src/java.base/share/classes/jdk/internal/foreign/layout/LayoutTransformers.java

```diff
@@ -0,0 +1,261 @@
+/*
+ *  Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ *  DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ *  This code is free software; you can redistribute it and/or modify it
+ *  under the terms of the GNU General Public License version 2 only, as
+ *  published by the Free Software Foundation.  Oracle designates this
+ *  particular file as subject to the "Classpath" exception as provided
+ *  by Oracle in the LICENSE file that accompanied this code.
+ *
+ *  This code is distributed in the hope that it will be useful, but WITHOUT
+ *  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ *  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ *  version 2 for more details (a copy is included in the LICENSE file that
+ *  accompanied this code).
+ *
+ *  You should have received a copy of the GNU General Public License version
+ *  2 along with this work; if not, write to the Free Software Foundation,
+ *  Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ *   Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ *  or visit www.oracle.com if you need additional information or have any
+ *  questions.
+ *
+ */
+
+package jdk.internal.foreign.layout;
+
+import jdk.internal.ValueBased;
+
+import java.lang.foreign.AddressLayout;
+import java.lang.foreign.GroupLayout;
+import java.lang.foreign.MemoryLayout;
+import java.lang.foreign.PaddingLayout;
+import java.lang.foreign.SequenceLayout;
+import java.lang.foreign.StructLayout;
+import java.lang.foreign.UnionLayout;
+import java.lang.foreign.ValueLayout;
+import java.nio.ByteOrder;
+import java.util.Objects;
+import java.util.function.Function;
+
+/**
+ * A JDK-internal utility class that provides useful layout transformations.
+ */
+public final class LayoutTransformers {
```

Comment:
In my experience when thinking about this problem, I always found that writing a layout transformer as a method implemented by pattern matching is almost always a better choice. The problem with going down that path is that it's a little verbose in cases where no transform needs to occur. Let's consider the endianness case:

```
MemoryLayout setEndianness(MemoryLayout layout, ByteOrder order) {
     return switch (layout) {
           case ValueLayout vl -> yield vl.order(order);
           case SequenceLayout sl -> yield MemoryLayout.sequenceLayout(sl.elementCount(), setEndianness(sl.elementLayout(), order);
           case StructLayout stl -> yield MemoryLayout.structLayout(stl.memberLayouts().stream().map(e -> setEndianness(e, order));
           ...
}
```

The problem with this is that the arms where "nothing" happens need to draw a lot of attention, as they need to contain the code which recursively propagates the transform.

But what if we had a single utility like this:

```
MemoryLayout adaptContents(MemoryLayout, UnaryOperator<MemoryLayout>)
```

Which takes care of applying a transform recursively on the _contents_ of a nested layout (e.g. sequence layout element, or the members of a group layout). Then the above becomes:

```
MemoryLayout setEndianness(MemoryLayout layout, ByteOrder order) {
     return switch (layout) {
           case ValueLayout vl -> yield vl.order(order);
           default -> yield layout.adaptContents(e -> setEndianness(e, order));
     }
}
```

Which is quite simple.

And, the remove names transform can be even simpler:

```
MemoryLayout removeNames(MemoryLayout layout) {
     return layout.adaptContents(this::removeNames).withoutName();
}
```